Update lambdas.md

Resuscitating opened the pull request:
I found your guide here, and wanted to contribute. What I've written here is more of a stub than a full page, so I will continue to work on it throughout the next few days. I've done my best to emulate your style of writing, but I'm not very good at imitating other people or writing overall, so I fear that it won't be very good. If you have any feedback or see anything I can do differently/better, please let me know.

Commit:
Update lambdas.md

I found your guide here, and wanted to contribute. What I've written here is more of a stub than a full page, so I will continue to work on it throughout the next few days. I've done my best to emulate your style of writing, but I'm not very good at imitating other people or writing overall, so I fear that it won't be very good. If you have any feedback or see anything I can do differently/better, please let me know.

IFcoltransG commented:
I'm glad you're interested.
I've looked at my notebook for what I thought would be good to cover in the lambdas sections, and I've written "decorators, functions, exprs". I think my plan was to start with talking about functions (possibly including code objects but that could easily go somewhere else) then move into lambdas, which would naturally lead to talking about expressions versus statements. I can provide some examples of fun lambdas that do things one would only expect from statements.
As for the writing style, I don't recommend trying to be me. You've got your own voice, I'm certain. The point of this guide isn't to sound homogenous, it's to inform.
But I'll give some context for how I was writing. My rules of thumb are, it should convey the information, and it should keep things lively. I myself am drawing a lot from Learn You A Haskell when I write the guide. Jokes and informal language are well and good, but I also tried to break things up with formatting. Formatting includes text styling and headings. Code blocks add splashes of colour, so I try to thread short ones in with the monologue. If I can't think any witty joke, I just write as though I'm having a great time. That tends to make things more lively. The target audience I imagine in my mind is someone who has seen and written a fair amount of regular Python and knows *what* things do, but not *why* everything works. All this writing does take a while to put together me, hence it's fallen to the wayside a bit as I do other things. (The liveliness part is not my normal register I write in.)
I'll probably make lots of suggestions about the content, but I'm happy to be as light- or heavy-handed as you'd like for the voice/style. If I don't change much about the voice then it should be advertised as a 'guest article' or the like, because it would be in your words. You'd be in the credits either way.

IFcoltransG commented on src/lambdas.md:
````diff
@@ -1,2 +1,26 @@
+Given that you're reading a guide for writing esoteric Python, you probably know what functions are. The typical function is something like this:
+```python
+def foo(spam, eggs):
+    # function body here
````
If this function had a return value, it would be easier to see a direct equivalence with lambda functions.

IFcoltransG commented on src/lambdas.md:
````diff
@@ -1,2 +1,26 @@
+## Syntax
+Unlike a regular function, the entirety of a lambda function declaration is an expression. This means that a lambda can be declared inline. A lambda declaration is as follows:
+```python
+lambda spam, eggs: return_value
````
At some point it would be good to have talked about functions as first class citizens. That is, functions are just variables like anything else. If you explain that `def foo()` assigns to `foo`, then that might make it clearer what a lambda is meant to represent.
There are two things standing in the way of learning what a lambda is: fear ("lambda is a weird word I've never seen before, and I've heard they're complicated") and the unfamiliarity of using functions as first class objects in Python.

IFcoltransG commented on src/lambdas.md:
````diff
@@ -1,2 +1,26 @@
+```python
+lambda spam, eggs: return_value
+```
+This resembles a normal function declaration, but there are a few differences; `def` is replaced with `lambda`, there is no function name, and there are no parentheses around the argument listing. Because a lambda declaration is an expression, you don't use the `return` keyword to return a value. Instead, the expression after the argument listing is what is evaluated and returned upon calling the ...
````
> no function name

If you introduced lambdas similarly to the Python docs (from memory) as `foo = lambda:` then it might be easier to see the relationship to Naveed functions, and wean readers onto using a lambda without a name.

IFcoltransG commented on src/lambdas.md:
```diff
@@ -1,2 +1,26 @@
+This resembles a normal function declaration, but there are a few differences; `def` is replaced with `lambda`, there is no function name, and there are no parentheses around the argument listing. Because a lambda declaration is an expression, you don't use the `return` keyword to return a value. Instead, the expression after the argument listing is what is evaluated and returned upon calling the ...
+
+## Usage in esoteric python
+Because lambdas are expressions, you can use them immediately after declaration, instead of having to assign to an intermediate variable. For example:
```
You'll be able to directly compare how you might use a function to how you might use a lambda.
Also if you'll use `map` it could be a good time to explain what the function does.

IFcoltransG commented on src/lambdas.md:
```diff
@@ -1,2 +1,26 @@
 # Lambdas
```
The title doesn't need to be taken literally. As I outlined in the PR comment, I thought this section could also include general information about functions if you wanted.

IFcoltransG replied:
In other sections I've gone on tangents. It's my habit to want to cover everything, I only restrain myself by planning to put things later on.

Resuscitating replied:
> The title doesn't need to be taken literally. As I outlined in the PR comment, I thought this section could also include general information about functions if you wanted.

My current draft of the section is more about functions in general than lambdas specifically. I have a decently sized chunk about decorators, which I plan on going relatively in-depth in, and then I'm going to talk about lambdas.

IFcoltransG commented:
Also I should confirm, are you okay with your contribution being licensed (like the rest of this repo) under the [Unlicense](https://unlicense.org/), i.e. completely public domain and waiving copyright?

ghost commented:
Thanks for the feedback on my writing! I'll work on adding more detail about "decorators, functions, exprs" and the like. As far as providing examples, that's probably not necessary. I can make/come up with my own.

As for the license, I'm completely fine with that. I have no qualms about the work being public domain.

IFcoltransG commented:
Great to hear. I've left some comments on specific lines of the pull request as well.

ghost commented:
> Great to hear. I've left some comments on specific lines of the pull request as well.

How do I view those? I'm relatively new to github so I don't know my way around yet.

IFcoltransG commented:
It turns out it was my fault, I hadn't confirmed my review. They should be readable now if you go to files changed.

Commit:
First Draft

Still needs more details about decorators and lambdas, but I think it's a good first draft. I'll continue to work on it when I can.
